File: c/queue/array_queue.c
#include <stdio.h>
#include <stdbool.h>
#include <stdlib.h>

struct array_queue {
    int* data;
    int size;
    int head;
    int tail;
};

bool init_queue(struct array_queue* queue, int size) {
    queue->data = malloc(sizeof(int) * size);

    if (queue->data == NULL) {
        printf("data malloc error\n");
        return false;
    }

    queue->size = size;
    queue->head = 0;
    queue->tail = 0;

    return true;
}

bool enqueue(struct array_queue* queue, int data) {

    if (queue->tail == queue->size) {
        if (queue->head == 0) {
            printf("    ->queue is full, enqueue fail\n");
            return false;
        }

        for (int i = 0; i < queue->tail; i++) {
            queue->data[i - queue->head] = queue->data[i];
        }

        queue->tail = queue->tail - queue->head;
        queue->head = 0;
    }

    queue->data[queue->tail++] = data;

    return true;
}

int dequeue(struct array_queue* queue) {
    if (queue->head == queue->tail) {
        printf("    ->queue is empty, dequeue fail\n");
        return -1;
    }

    return queue->data[queue->head++];
}


int main(void) {

    struct array_queue queue;
    
    init_queue(&queue, 3);

    enqueue(&queue, 1);
    enqueue(&queue, 2);
    enqueue(&queue, 3);
    enqueue(&queue, 4);

    printf("data : %d\n", dequeue(&queue));
    printf("data : %d\n", dequeue(&queue));
    printf("data : %d\n", dequeue(&queue));
    printf("data : %d\n", dequeue(&queue));

    enqueue(&queue, 5);
    printf("data : %d\n", dequeue(&queue));
    printf("data : %d\n", dequeue(&queue));


    return 1;
}

File: python/queue/array_queue.py

class ArrayQueue(object):
    def __init__(self, size):
        self.data = [None] * size
        self.size = size
        self.head = 0
        self.tail = 0

    def equeue(self, data : int) -> bool:

        if self.tail == self.size:
            if self.head == 0:      # queue is really full
                print('    -> queue is full')
                return False

            for i in range(self.head, self.tail):       # copy data when no space at tail
                self.data[i - self.head] = self.data[i]

            self.tail = self.tail - self.head
            self.head = 0

        self.data[self.tail] = data
        self.tail += 1

        return True

    def dequeue(self) -> int:

        if self.head == self.tail:
            print('    -> queue is empty, dequeue fail')
            return None

        data = self.data[self.head]
        self.head += 1

        return data
        

def main():
    queue = ArrayQueue(3)

    queue.equeue(1)
    queue.equeue(2)
    queue.equeue(3)
    queue.equeue(4)

    print('data : %s' % queue.dequeue())
    print('data : %s' % queue.dequeue())
    print('data : %s' % queue.dequeue())
    print('data : %s' % queue.dequeue())

if __name__ == '__main__':
    main()